webroot/js/app.js

```diff
@@ -109,53 +127,17 @@ class Owncast {
 
   // websocket for messaging
   setupWebsocket() {
-    var ws = new WebSocket(URL_WEBSOCKET);  
-    ws.onopen = (e) => {
-      if (this.websocketReconnectTimer) {
-        clearTimeout(this.websocketReconnectTimer);
-      }
-
-      // If we're "online" then enable the chat.
-      if (this.streamStatus && this.streamStatus.online) {
-        this.messagingInterface.enableChat();
-      }
-    };
-    ws.onclose = (e) => {
-      // connection closed, discard old websocket and create a new one in 5s
-      this.websocket = null;
-      this.messagingInterface.disableChat();
-      this.handleNetworkingError('Websocket closed.');
-      this.websocketReconnectTimer = setTimeout(this.setupWebsocket, TIMER_WEBSOCKET_RECONNECT);
-    };
-    // On ws error just close the socket and let it re-connect again for now.
-    ws.onerror = e => {
-      this.handleNetworkingError(`Socket error: ${JSON.parse(e)}`);
-      ws.close();
-    };
-    ws.onmessage = (e) => {
-      const model = JSON.parse(e.data);
-
-      // Send PONGs
-      if (model.type === SOCKET_MESSAGE_TYPES.PING) {
-        this.sendPong(ws);
-        return;
-      } else if (model.type === SOCKET_MESSAGE_TYPES.CHAT) {
-        const message = new Message(model);
-        this.addMessage(message);
-      } else if (model.type === SOCKET_MESSAGE_TYPES.NAME_CHANGE) {
-        this.addMessage(model);
-      }
-    };
-    this.websocket = ws;
-    this.messagingInterface.setWebsocket(this.websocket);
+    this.websocket = new Websocket();
+    this.websocket.addListener(this);
```

Comment:
Hm interesting. I was wondering how "this" whole thing was a "listener". But I guess this works.

Reply:
This still feels a bit strange to me, maybe syntactically.. 
Could we instead do something more explicit and in a typical format such as: `this.websocket.addListener(CALLBACKS.WEBSOCKET_DISCONNECTED, this.someCallback)`.

Reply:
I know it's a little more wordy, but I think it would help to read the code better and be able to see where some of the methods are going.

Reply:
We could, that would be more like Javascript events?  And I guess we could just use random event emitters at that point, but those make me cringe, but that's the javascripty way of doing it I suppose.  I wanted to build a standardized interface that any object could implement, it is much cleaner in my eyes since all objects will follow the same rules, standardizing the names of things.  If you're an object and you want to know when the websocket connects you implement `websocketConnected`, always.  I prefer this over some objects having `websocketConnected` and others having `socketOpened` and others having `chatStarted`.

Though I admit I don't think Javscript does interfaces and/or protocols, but that's what I'm trying to do here.  A shared contract between the websocket library and those that use it.  But if that's confusing I can change it.

Reply:
Well since it's still all javascript here, structure-wise it's much more understandable. We don't have to use actual cringey event emitters. The only events happening are the websocket's events. The standard between each new class that uses this Websocket class would only be being aware of its CALLBACKS.[event name]s, which makes sense to me and makes it more easy to remember to implement a callback for.

Reply:
Ok so how about this, I'll do the Javascripty thing and use events.

Two changes:
* `Websocket` will no longer call callbacks, but instead emit events on the listeners.
```
const event = new Event('websocketConnected');
listener.dispatchEvent(event)
```

And listeners will no longer implement the required interface functions, but instead just listen for events and run whatever it feels like.

```
this.addEventListener('websocketConnected', function (event) {});
```

Or whatever the syntaxes are.  Isn't there some kind of `.on('event')` syntax too?

Reply:
Oh I was saying we didn't have emit events.

Reply:
Seems like the `on` syntax is a jQuery thing.  The benefit I see over events (even though they're ugly) is `App` won't have to know the inner workings of `MessagingInterface`.  Otherwise if we explicitly make `App` set callbacks for `MesagingInterface` then `App` has to know methods inside `MesagingInterface`, and that feels leaky.

Reply:
Ok updates with the
```
this.websocket.addListener('rawWebsocketMessageReceived', this.receivedWebsocketMessage.bind(this));
```
Approach is in!

Reply:
Looks good to me 👍